lib/pure/asyncdispatch.nim

```diff
@@ -355,6 +355,28 @@ proc `or`*[T, Y](fut1: Future[T], fut2: Future[Y]): Future[void] =
   fut2.callback = cb
   return retFuture
 
+proc all*[A](futs: seq[Future[A]]): Future[seq[A]] =
+  ## Returns a future which will complete once all futures in ``futs``
+  ## complete.
+  ##
+  ## The resulting future will hold the values of all awaited futures,
+  ## in the order they are passed.
+
+  var
+    retFuture = newFuture[seq[A]]("asyncdispatch.all")
+    retValues = newSeq[A](len(futs))
+    completedFutures = 0
+
+  for i, fut in futs:
+    fut.callback = proc(f: Future[A]) =
+      retValues[i] = f.read()
```

Comment:
How about another closure instead of ctx?
```nim
proc all*[A](futs: openarray[Future[A]]): Future[seq[A]] =
  ## Returns a future which will complete once all futures in ``futs``
  ## complete.
  ##
  ## The resulting future will hold the values of all awaited futures,
  ## in the order they are passed.

  var
    retFuture = newFuture[seq[A]]("asyncdispatch.all")
    retValues = newSeq[A](len(futs))
    completedFutures = 0

  for i, fut in futs:
    proc setCb(i: int) =
      fut.callback = proc(f: Future[A]) =
        retValues[i] = f.read()
        inc(completedFutures)

        if completedFutures == len(retValues):
          retFuture.complete(retValues)
    setCb(i)

  return retFuture
```